gist.go

```diff
@@ -0,0 +1,92 @@
+package model
+
+// Typed constants. Types are not exported as you can still export the values 
+// for use in `if` and `switch` statements.
+type channelType string
+
+// CamelCase Constant Names
+// Constants are exported for use in `if` and `switch` statements and similar in other packages.
+const (
+	ChannelTypeOpen    channelType = "O"
+	ChannelTypePrivate channelType = "P"
+)
+
+type Channel struct {
+	// Member variables are not exported to ensure immutability.
+	id       string `json:"id"`
+	name     string `json:"name"`
+	createAt int64  `json:"create_at"`
+	// etc...
+
+	// Question: what do we do when there's a semantically meaningful
+	// "null" value for a member variable?
+	schemeID *string `json:"scheme_id"`
+
+	// Question: is this the best way of storing props and similar map/list data?
+	props map[string]interface{} `json:"props" db:"-"`
+}
+
+// Getters to access the member variables are provided. These *never* return pointers.
+// Correct capitalisation of initialisms
+func (channel *Channel) ID() string {
+	return channel.id
+}
+
+func (channel *Channel) Name() string {
+	return channel.name
+}
+
+// Returning `interface{}` feels gross. Can we do better than this?
+func (channel *Channel) GetProp(key string) interface{} {
+	return channel.props[key]
+}
+
+// "Patch" object holds changes to the object to make.
+type ChannelPatch struct {
+	Name *string `json:"name"`
+	// etc...
+
+	// Question: can we avoid pointers here and have some type like "NullableString"
+	// that indicates whether the field should be set explicitly instead?
```

Comment:
@agnivade any thoughts on this? I can't decide whether I just want to universally do away with pointer types for struct fields, or whether the use of pointers here is reasonable.

Reply:
I was just talking about this to @isacikgoz and I believe we have found a decent solution to this problem. The idea is to use default values in struct tags, and use reflection to figure out whether the field was explicitly set or not. It's something that is used in protobuf too, and I think it can solve a lot of our problems including removing _all_ pointer fields in our Config struct, and get rid of the madness that is `NewBool`, `NewInt`.

It also has some added advantages like setting default values side by side with the field declaration, so that the developer doesn't have to declare yet another constant and write 3 more lines to check for a nil field and if so, then set the pointer.

It's still at the research stage, so no promises :smile:. But @isacikgoz is working on it. If we see good results, then we can reserve some mana to refactor that logic to a separate repo and completely revamp this. Also a bonus point, this also takes care of validating fields. So there will be no need for `IsValid` methods any more.

Reply:
@agnivade Could you elaborate a bit how struct tags can be used for validation? Did you had using https://github.com/go-playground/validator in mind?

Reply:
Something like that, yes. More info here: https://community-daily.mattermost.com/core/channels/developers-server/dxjf4fusxfrh9gr4fqgt79gafy. Please feel free to join today's server call.